src/accountsdb/db.zig

```diff
@@ -808,263 +997,1125 @@ pub const AccountsDB = struct {
+    /// as the data field ([]u8) for each account.
+    pub fn flushSlot(self: *Self, slot: Slot, unclean_account_files: *ArrayList(FileId)) !void {
+        defer self.stats.number_files_flushed.inc();
 
-    pub inline fn slotListMaxWithinBounds(
-        ref_ptr: *AccountRef,
-        min_slot: ?Slot,
-        max_slot: ?Slot,
-    ) ?*AccountRef {
-        var biggest: ?*AccountRef = null;
-        if (inBoundsIf(ref_ptr.slot, min_slot, max_slot)) {
-            biggest = ref_ptr;
-        }
+        const pubkeys, const accounts: []Account = blk: {
+            // NOTE: flush should the only function to delete/free cache slices of a flushed slot
+            // -- purgeSlot removes slices but we should never purge rooted slots
+            const account_cache, var account_cache_lg = self.account_cache.readWithLock();
+            defer account_cache_lg.unlock();
 
-        var curr = ref_ptr;
-        while (curr.next_ptr) |ref| {
-            if (inBoundsIf(ref.slot, min_slot, max_slot) and (biggest == null or ref.slot > biggest.?.slot)) {
-                biggest = ref;
-            }
-            curr = ref;
+            const pubkeys, const accounts = account_cache.get(slot) orelse return error.SlotNotFound;
+            break :blk .{ pubkeys, accounts };
+        };
+        std.debug.assert(accounts.len == pubkeys.len);
+
+        // create account file which is big enough
+        var size: usize = 0;
+        for (accounts) |*account| {
+            size += account.getSizeInFile();
         }
-        return biggest;
-    }
+        const file, const file_id, const memory = try self.createAccountFile(size, slot);
 
-    pub fn getAccountFromRef(self: *const Self, account_ref: *const AccountRef) !Account {
-        switch (account_ref.location) {
-            .File => |ref_info| {
-                const account_in_file = try self.storage.getAccountInFile(
-                    ref_info.file_id,
-                    ref_info.offset,
-                );
-                const account = Account{
-                    .data = account_in_file.data,
-                    .executable = account_in_file.executable().*,
-                    .lamports = account_in_file.lamports().*,
-                    .owner = account_in_file.owner().*,
-                    .rent_epoch = account_in_file.rent_epoch().*,
-                };
-                return account;
-            },
-            .Cache => |ref_info| {
-                const account = self.storage.cache.items[ref_info.index];
-                return account;
-            },
+        const offsets = try self.allocator.alloc(u64, accounts.len);
+        defer self.allocator.free(offsets);
+        var offset: usize = 0;
+        for (0..accounts.len) |i| {
+            offsets[i] = offset;
+            // write the account to the file
+            offset += accounts[i].writeToBuf(&pubkeys[i], memory[offset..]);
         }
-    }
+        try file.sync();
```

Comment:
Is this necessary? I thought that as soon as you write it to the mmapped memory, that writes it to the virtual filesystem that the os exposes to all processes. In other words, any read of this file will see the data that you've written, regardless of it actually being on the disk. I thought manual syncing is only necessary if you expect the drive to suddenly get disconnected, otherwise for all intents and purposes it behaves as if the data is on the disk. I may have a misconception here but I could have sworn that's how it works.

Reply:
yeah youre right - the original rust implementation used sync which is why it was there but i saw a recent pr  (missing the link unfortunately) saying the same as what you mentioned and removing it

7afd16a